### ISSUE TITLE
[analyzer] Implement FutureOr related rules of least upper bound

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#43720]
This is the analyzer specific issue for #43719, which has the details.

**Comment by scheglov** [comment]
The test below fails because `UP(FutureOr<void>, dynamic)` is `dynamic`. We apply `UP(T1, T2) where TOP(T1) and TOP(T2)` and `MORETOP(FutureOr<void>, dynamic)` is `false`, so the result is `T2` which is `dynamic`.

```dart
  void f20(dynamic a, FutureOr<void> b) {
    var x = condition ? a : b;
    x.expectStaticType<Exactly<FutureOr<void>>>();
    if (x is Future) throw 0;
    /**/ x.toString();
    //   ^
    // [analyzer] COMPILE_TIME_ERROR.USE_OF_VOID_RESULT
    // [cfe] This expression has type 'void' and can't be used.

    var y = condition ? b : a;
    y.expectStaticType<Exactly<FutureOr<void>>>();
    if (y is Future) throw 0;
    /**/ y.toString();
    //   ^
    // [analyzer] COMPILE_TIME_ERROR.USE_OF_VOID_RESULT
    // [cfe] This expression has type 'void' and can't be used.
  }
```

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/166786

**Comment by eernstg** [comment]
@scheglov wrote:
> .. the result is `T2` which is `dynamic`.

Indeed, thanks for catching that! Fixed in dart-lang/sdk@a369992f4bf106f127d530db43cb11e9228ac4e8.
